File: savewise-client/src/services/category-service.ts
import { CategoriesResponse, Category, CategoryResponse, CategoryTypesResponse } from "../common/objects/categories";
import { Status } from "../common/objects/response";

const apiUrl = process.env.REACT_APP_API_BASE_URL;
const baseUrl = `${apiUrl}/categories`;

export interface GetCategoriesInput {
  includeAmounts: boolean,
  startDate: string,
  endDate: string,
  categoryTypeId?: number
}

export function getCategories(userId: number, searchOptions: GetCategoriesInput): Promise<CategoriesResponse> {
    if (!userId) return Promise.reject("User id null");
    const url = `${baseUrl}/user/${userId}`;
    const body = JSON.stringify(searchOptions);
    const options: RequestInit = {
      method: 'POST',
      mode: 'cors',
      headers: {
        'Content-Type': 'application/json',    
        'Access-Control-Allow-Origin':'*',
      },
      body
    };
    return fetch(url, options).then(response =>
      response.json()
    );
  }

export function getCategoryTypes(): Promise<CategoryTypesResponse> {
    const url = `${baseUrl}/types`;
    return fetch(url).then(response =>
      response.json()
    );
  }

export function saveCategory(category: Category): Promise<CategoryResponse> {
  if (!category) return Promise.reject("Category null");
    const url = baseUrl;
    const body = JSON.stringify(category);
    const options: RequestInit = {
      method: 'POST',
      mode: 'cors',
      headers: {
        'Content-Type': 'application/json',    
        'Access-Control-Allow-Origin':'*',
      },
      body
    };
    return fetch(url, options).then(r => r.json());
}

export function editCategory(category: Category): Promise<CategoryResponse> {
  if (!category || !category.id) return Promise.reject("Category or category.id null");
  const url = `${baseUrl}/${category.id}`;
  const body = JSON.stringify(category);
  const options: RequestInit = {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',    
      'Access-Control-Allow-Origin':'*',
    },
    body
  };
  return fetch(url, options).then(r => r.json());
}

export function deleteCategory(categoryId: number): Promise<Status> {
  if (!categoryId) return Promise.reject("Category id null");
  const url = `${baseUrl}/${categoryId}`;
  const options: RequestInit = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json',    
      'Access-Control-Allow-Origin':'*',
    }
  };
  return fetch(url, options).then(r => r.json());
}

File: SavewiseAPI/Services/CategoryService.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Savewise.Managers;
using Savewise.Models;
using Savewise.Services;
using Savewise.Services.Objects;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Savewise.Services
{
    [Route("api/categories")]
    public class CategoryService : BaseService
    {
        public class CategoryResponse : ServiceResponse
        {
            public OCategory category { get; set; }
        }
        public class CategoriesResponse : ServiceResponse
        {
            public List<OCategory> categories { get; set; }
        }
        public class CategoryTypesResponse : ServiceResponse
        {
            public List<OCategoryType> categoryTypes { get; set; }
        }

        public class GetCategoriesInput
        {
            public bool includeAmounts { get; set; }
            public string startDate { get; set; }
            public string endDate { get; set; }
            public int? categoryTypeId { get; set; }
        }
        public CategoryService(SavewiseContext context): base(context)
        {

        }

        // POST api/categories/user/{id}
        [HttpPost("user/{id}")]
        public IActionResult GetAll(int id, [FromBody] GetCategoriesInput input)
        {
            CategoriesResponse response = new CategoriesResponse();
            response.status = new Status();
            response.status.success = false;
            try
            {
                CategoryManager manager = new CategoryManager(context);
                response.categories = manager.getAll(id, input.includeAmounts, input.startDate, input.endDate, input.categoryTypeId);
                response.status.success = true;
            }
            catch (Exception exception)
            {
                response.status.errorMessage = exception.Message;
            }
            return Json(response);
        }

        // GET api/categories/{id}
        [HttpGet("{id}")]
        public IActionResult GetCategory(int id)
        {
            CategoryResponse response = new CategoryResponse();
            response.status = new Status();
            response.status.success = false;
            try
            {
                CategoryManager manager = new CategoryManager(context);
                response.category = manager.getById(id);
                response.status.success = true;
            }
            catch (Exception exception)
            {
                response.status.errorMessage = exception.Message;
            }
            return Json(response);
        }

        // GET api/categories/types
        [HttpGet("types")]
        public IActionResult GetCategoryTypes()
        {
            CategoryTypesResponse response = new CategoryTypesResponse();
            response.status = new Status();
            response.status.success = false;
            try
            {
                CategoryManager manager = new CategoryManager(context);
                response.categoryTypes = manager.GetCategoryTypes();
                response.status.success = true;
            }
            catch (Exception exception)
            {
                response.status.errorMessage = exception.Message;
            }
            return Json(response);
        }

        // POST api/categories/
        [HttpPost()]
        public IActionResult SaveNew([FromBody] OCategory category)
        {
            CategoryResponse response = new CategoryResponse();
            response.status = new Status();
            response.status.success = false;
            try
            {
                using (var transaction = context.Database.BeginTransaction())
                {
                    CategoryManager manager = new CategoryManager(context);
                    response.category = manager.save(category);
                    response.status.success = true;

                    transaction.Commit();
                }
            }
            catch (Exception exception)
            {
                response.status.errorMessage = exception.Message;
            }
            return Json(response);
        }

        // PATCH api/categories/{id}
        [HttpPatch("{id}")]
        public IActionResult EditCategory([FromBody] OCategory category)
        {
            CategoryResponse response = new CategoryResponse();
            response.status = new Status();
            response.status.success = false;
            try
            {
                using (var transaction = context.Database.BeginTransaction())
                {
                    CategoryManager manager = new CategoryManager(context);
                    response.category = manager.save(category);
                    response.status.success = true;

                    transaction.Commit();
                }
            }
            catch (Exception exception)
            {
                response.status.errorMessage = exception.Message;
            }
            return Json(response);
        }

        // DELETE api/categories/{id}
        [HttpDelete("{id}")]
        public IActionResult DeleteCategory(int id)
        {
            Status response = new Status();
            response.success = false;
            try
            {
                using (var transaction = context.Database.BeginTransaction())
                {
                    CategoryManager manager = new CategoryManager(context);
                    response.success = manager.delete(id);

                    transaction.Commit();
                }
            }
            catch (Exception exception)
            {
                response.errorMessage = exception.Message;
            }
            return Json(response);
        }
    }
}
